text/0000-driver-model.md

```diff
@@ -0,0 +1,328 @@
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The twizzler-driver crate provides a high-level wrapper around a collection of abstractions designed
+to manage a single device (for definition of "device" and the `Device` type, see RFCxxxx), called a
+`DeviceController`. When taking control of a device, driver software can create a `DeviceController`
+(heretoafter referred to as "the controller") from a `Device`. After this point, the controller
+manages the device and exposes several abstractions above the device:
+
+ - A `DeviceEventStream`, which provides a way to get a stream of events, including mailbox events
+   and device interrupts (see RFCxxxx).
+ - A `DmaAllocator` (details for which will appear in a future RFC).
+ - A `RequestEngineManager`, which provides a way to submit requests to the device and
+   asynchronously await responses.
+
+## Device Event Stream
+
+The device event stream provides an async interface for handling (higher-half) interrupts and
+mailbox messages. This means that some async executor is required; if this is too much overhead for
+the application, direct access to the underlying Device allows for lower-level interrupt handling
+without the async overhead.
+
+This allows driver software to handle incoming events from the device. For example, a NIC that
+receives a packet transfers the packet data to a receive buffer and then submits an interrupt.
+Driver software will then see this interrupt via the event stream and can handle it appropriately.
+
+The `DeviceEventStream` provides an interface for allocating an interrupt, which return an
+`InterruptInfo`. Internally, the twizzler-driver crate handles the allocation of the interrupt
+according to the transport mechanism (e.g. PCIe MSI/MSI-X) and registration with the kernel. The
+`InterruptInfo` type exposes an async function `next` which returns the non-zero u64 value of the
+next interrupt that fires or None if the interrupt is removed. Thus driver software can handle a
+stream of interrupts as follows:
+
+```{rust}
+let int = controller.events().allocate_interrupt(...).unwrap();
+while let Some(x) = int.next().await {
+    while still_work_to_do() {
+        // handle interrupt
+    }
+}
+```
+
+The `DeviceEventStream` also provides a method for accessing mailboxes, which can be used as
+follows:
+
+```{rust}
+loop {
+    handle_mailbox_message(controller.events().next_msg(MailboxPriority::High).await);
+}
+```
+
+As in RFCxxxx, it is up to driver software to ensure correct prioritization of mailbox receivers.
+
+## Request Engine Manager
+
+The other main abstraction is to model the device such that it receives requests and asynchronously
+responds. For example, submitting a packet for transmit to a NIC happens by the driver software
+constructing an entry in a transmit queue and then notifying the device that the queue has been
+appended to. After the device submits the packet, it responds by indicating that that queue entry
```

Comment:
In most devices this is a ring, not a queue.  Yes, we call it the transmit queue but...I think it depends on how faithful to "real" hardware we want to be in this discussion.